File: MuseoPOO/catalogo.py
"""Módulo que define la clase Catalogo."""

from __future__ import annotations

from typing import TYPE_CHECKING, Optional

if TYPE_CHECKING:
    from obra_de_arte import ObraDeArte


class Catalogo:
    """Representa el catálogo de obras del museo."""

    def __init__(self) -> None:
        self._obras: list[ObraDeArte] = []

    def get_obras(self) -> list[ObraDeArte]:
        return list(self._obras)

    def agregar_obra(self, obra: ObraDeArte) -> None:
        """Agrega una obra al catálogo."""
        if obra not in self._obras:
            self._obras.append(obra)
            print(
                f"Obra '{obra.get_titulo()}' registrada "
                f"en el catálogo."
            )
        else:
            print(
                f"La obra '{obra.get_titulo()}' ya está "
                f"en el catálogo."
            )

    def buscar_obra(self, titulo: str) -> Optional[ObraDeArte]:
        """Busca una obra en el catálogo por su título."""
        for obra in self._obras:
            if obra.get_titulo().lower() == titulo.lower():
                return obra
        print(
            f"Obra con título '{titulo}' no encontrada "
            f"en el catálogo."
        )
        return None

    def listar_obras(self) -> list[ObraDeArte]:
        """Retorna todas las obras del catálogo."""
        return list(self._obras)

    def __str__(self) -> str:
        return f"Catalogo(obras={len(self._obras)})"

File: Museo Java/Catalogo.java
import java.util.ArrayList;
import java.util.List;

public class Catalogo {

    private List<ObraDeArte> obras;

    public Catalogo() {
        this.obras = new ArrayList<>();
    }

    public List<ObraDeArte> getObras() { return new ArrayList<>(obras); }

    public void agregarObra(ObraDeArte obra) {
        if (!obras.contains(obra)) {
            obras.add(obra);
            System.out.println("Obra '" + obra.getTitulo() + "' registrada en el catálogo.");
        } else {
            System.out.println("La obra '" + obra.getTitulo() + "' ya está en el catálogo.");
        }
    }

    public ObraDeArte buscarObra(String titulo) {
        for (ObraDeArte obra : obras) {
            if (obra.getTitulo().equalsIgnoreCase(titulo)) {
                return obra;
            }
        }
        System.out.println("Obra con título '" + titulo + "' no encontrada en el catálogo.");
        return null;
    }

    public List<ObraDeArte> listarObras() {
        return new ArrayList<>(obras);
    }

    @Override
    public String toString() {
        return "Catalogo(obras=" + obras.size() + ")";
    }
}
